patches/tModLoader/Terraria/ModLoader/ModEmoteBubble.cs

```diff
@@ -0,0 +1,153 @@
+﻿using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using Terraria.GameContent.UI;
+using Terraria.GameContent.UI.Elements;
+using Terraria.Localization;
+
+namespace Terraria.ModLoader
+{
+	public abstract class ModEmoteBubble : ModType<EmoteBubble, ModEmoteBubble>
+	{
+		/// <summary>
+		/// The file name of this emote's texture file in the mod loader's file space.
+		/// </summary>
+		public virtual string Texture => (GetType().Namespace + "." + Name).Replace('.', '/');
+
+		/// <summary>
+		/// The internal ID of this EmoteBubble.
+		/// </summary>
+		public int Type { get; internal set; }
+
+		/// <summary>
+		/// This is the <see cref="EmoteBubble"/> instance.
+		/// </summary>
+		public EmoteBubble EmoteBubble { get; internal set; }
+
+		/// <summary>
+		/// This is the translation that is used behind <see cref="DisplayName"/>. The translation will show up as the emote command.
+		/// </summary>
+		public ModTranslation EmoteName { get; internal set; }
+
+		/// <summary>
+		/// This is the name that will show up as the emote command.
+		/// </summary>
+		public string DisplayName => EmoteName.GetTranslation(Language.ActiveCulture);
+
+		public sealed override void SetupContent() {
+			ModContent.Request<Texture2D>(Texture);
+			SetStaticDefaults();
+
+			if (EmoteName.IsDefault())
+				EmoteName.SetDefault(Name.ToLower()); // It will be the lowercase name of the class by default.
+		}
+
+		protected sealed override void Register() {
+			EmoteName = LocalizationLoader.GetOrCreateTranslation(Mod, $"EmoteName.{Name}");
+
+			ModTypeLookup<ModEmoteBubble>.Register(this);
+
+			Type = EmoteBubbleLoader.Add(this);
+		}
+
+		protected override EmoteBubble CreateTemplateEntity() => new(Type, new WorldUIAnchor()) { ModEmoteBubble = this };
+
+		/// <summary>
+		/// Allows you to add this emote to a specific vanilla category.
+		/// <br><b>This should only be called in <see cref="ModType.SetStaticDefaults"/></b></br>
+		/// </summary>
+		/// <param name="categoryId">The category to which this emote will be added. Use <see cref="EmoteID.Category"/> to select the category you want.</param>
+		public void AddToCategory(int categoryId) {
+			if (!EmoteBubbleLoader.categoryEmoteLookup.TryGetValue(categoryId, out var list)) {
+				EmoteBubbleLoader.categoryEmoteLookup.Add(categoryId, new() { this });
+				return;
+			}
+			list.Add(this);
+		}
+
+		/// <summary>
+		/// Allows you to determine whether or not this emote can be seen in emotes menu. Returns true by default.
+		/// <br/>Do note that this doesn't effect emote command and NPC using.
+		/// </summary>
+		/// <returns>If true, this emote will be shown in emotes menu.</returns>
+		public virtual bool IsUnlocked() => true;
+		
+		/// <summary>
+		/// Gets called when your emote bubble spawns in world.
+		/// </summary>
+		public virtual void OnSpawn() { }
```

Comment:
Should this be `OnSpawn`? Thoughts @JavidPack

Reply:
Sounds fine to me, other things that enter the world use OnSpawn